transformers4rec/tf/ranking_metric.py

```diff
@@ -84,54 +105,66 @@ def _metric(self, ks: tf.Tensor, preds: tf.Tensor, target: tf.Tensor) -> tf.Tens
 
 
 @ranking_metrics_registry.register_with_multiple_names("precision_at", "precision")
+@tf.keras.utils.register_keras_serializable(package="transformers4rec")
 class PrecisionAt(RankingMetric):
-    def __init__(self, top_ks=None, labels_onehot=False):
-        super(PrecisionAt, self).__init__(top_ks=top_ks, labels_onehot=labels_onehot)
+    def __init__(self, top_ks=None, labels_onehot=False, **kwargs):
+        super(PrecisionAt, self).__init__(top_ks=top_ks, labels_onehot=labels_onehot, **kwargs)
 
-    def _metric(self, ks: tf.Tensor, scores: tf.Tensor, labels: tf.Tensor) -> tf.Tensor:
+    def _metric(self, ks: List[int], scores: tf.Tensor, labels: tf.Tensor) -> tf.Tensor:
         """
         Compute precision@K for each provided cutoff in ks
-
         Parameters
         ----------
         {METRIC_PARAMETERS_DOCSTRING}
         """
 
         ks, scores, labels = check_inputs(ks, scores, labels)
         _, _, topk_labels = tf_utils.extract_topk(ks, scores, labels)
-        precisions = tf_utils.create_output_placeholder(scores, ks)
+        bs = tf.shape(scores)[0]
+
+        for index in range(int(tf.shape(ks)[0])):
+            k = ks[index]
+            rows_ids = tf.range(bs, dtype=tf.int64)
+            indices = tf.concat(
+                [
+                    tf.expand_dims(rows_ids, 1),
+                    tf.cast(index, tf.int64) * tf.ones([bs, 1], dtype=tf.int64),
+                ],
+                axis=1,
+            )
 
-        for index, k in enumerate(ks):
-            precisions[:, index].assign(tf.reduce_sum(topk_labels[:, : int(k)], axis=1) / float(k))
+            self.accumulator.scatter_nd_update(
+                indices=indices, updates=tf.reduce_sum(topk_labels[:, : int(k)], axis=1) / float(k)
+            )
 
-        return precisions
+        return self.accumulator
```

Comment:
Might be a little cleaner to have this method not return anything and just update `self.accumulator`. In `update_state` from the base-class we can just pass this variable on.

Reply:
yeah your right, I return it in the `_metric` because we are using this method as a nested call in `_metric` of AvgPrecisionAt  [here](https://github.com/NVIDIA-Merlin/Transformers4Rec/blob/c9aae3fedd1b1f71be6d93da3f4056808b7e27a4/transformers4rec/tf/ranking_metric.py#L220)  
Maybe we can have a flag in `_metric` to decide whether we want to return the tensor or not ?

Reply:
So the reason I brought this up was that implementing the `_metric` function is potentially user-code, since we want to allow users to create custom metrics. It would therefore be useful to make implementing this function as easy as possible. It's a bit tricky now that as a user now you have to know to update self.accumulator + returning it. Therefore I thought it would be simpler to just update the self.accumulator and we take care of the rest. We could create a public method that can be called if you want to call the _metric function and get the output back. What do you think?

Reply:
Thank you for the explanation! I totally agree with you. Regarding the name of the public method, can we call it something like `metric_fn`?

Reply:
I added the public method `metric_fn` for the cases where we need the output back and updated `_metric`.